src/java.base/share/classes/java/lang/invoke/LambdaForm.java

```diff
@@ -1374,49 +1358,40 @@ private Name(Name that, Object constraint) {
             this(new NamedFunction(function), arguments);
         }
         Name(NamedFunction function) {
-            this(-1, function.returnType(), function, EMPTY_ARGS);
+            this(-1, function.returnType(), function, EMPTY_ARGS, null);
         }
         Name(NamedFunction function, Object arg) {
-            this(-1, function.returnType(), function, new Object[] { arg });
+            this(-1, function.returnType(), function, new Object[] { arg }, null);
         }
         Name(NamedFunction function, Object arg0, Object arg1) {
-            this(-1, function.returnType(), function, new Object[] { arg0, arg1 });
+            this(-1, function.returnType(), function, new Object[] { arg0, arg1 }, null);
         }
         Name(NamedFunction function, Object... arguments) {
-            this(-1, function.returnType(), function, Arrays.copyOf(arguments, arguments.length, Object[].class));
+            this(-1, function.returnType(), function, Arrays.copyOf(arguments, arguments.length, Object[].class), null);
         }
         /** Create a raw parameter of the given type, with an expected index. */
         Name(int index, BasicType type) {
-            this(index, type, null, null);
+            this(index, type, null, null, null);
         }
         /** Create a raw parameter of the given type. */
         Name(BasicType type) { this(-1, type); }
 
         BasicType type() { return type; }
         int index() { return index; }
-        boolean initIndex(int i) {
-            if (index != i) {
-                if (index != -1)  return false;
-                index = (short)i;
-            }
-            return true;
-        }
+
         char typeChar() {
             return type.btChar;
         }
 
-        Name newIndex(int i) {
-            if (initIndex(i))  return this;
-            return cloneWithIndex(i);
-        }
-        Name cloneWithIndex(int i) {
-            Object[] newArguments = (arguments == null) ? null : arguments.clone();
-            return new Name(i, type, function, newArguments).withConstraint(constraint);
+        Name withIndex(int i) {
+            return new Name(i, type, function, arguments, constraint);
         }
```

Comment:
Maybe also do what `Name​::withConstraint​(Object)` does?
```suggestion
        Name withIndex(int i) {
            if (i == this.index) return this;
            return new Name(i, type, function, arguments, constraint);
        }
```

Reply:
Sure, initIndex and newIndex reuses same index object too.